app/vmselect/netstorage/netstorage.go

```diff
@@ -3286,6 +3270,31 @@ func (pnc *perNodeCounter) GetTotal() uint64 {
 // See https://github.com/golang/go/blob/704401ffa06c60e059c9e6e4048045b4ff42530a/src/runtime/malloc.go#L11
 const maxFastAllocBlockSize = 32 * 1024
 
+// execSearchQuery calls cb for with marshaled requestData for each tenant in sq.
+func execSearchQueryRequest(qt *querytracer.Tracer, sq *storage.SearchQuery, workerID uint, sn *storageNode, f func(mb *storage.MetricBlock, workerID uint) error, deadline searchutils.Deadline) error {
+	var requestData []byte
+
+	for i := range sq.TenantTokens {
+		requestData = sq.TenantTokens[i].Marshal(requestData)
+		requestData = sq.MarshalWithoutTenant(requestData)
+		qtL := qt
+		if sq.IsMultiTenant && qt.Enabled() {
+			qtL = qt.NewChild("query for tenant: %s", sq.TenantTokens[i].String())
+		}
+		sn.searchRequests.Inc()
+		if err := sn.processSearchQuery(qt, requestData, f, workerID, deadline); err != nil {
+			sn.searchErrors.Inc()
+			return fmt.Errorf("cannot perform search on vmstorage %s: %w", sn.connPool.Addr(), err)
+		}
```

Comment:
Seems like `qtL` should be used here. Also, if `err != nil` local qt will not be closed.